core/types/header_ext.go

```diff
@@ -0,0 +1,197 @@
+	out := new(HeaderSerializable)
+
+	updateFromEth(out, eth)
+	updateFromExtras(out, h)
+
+	return out.MarshalJSON()
+}
+
+//nolint:stdmethods
+func (h *HeaderExtra) UnmarshalJSON(eth *ethtypes.Header, input []byte) error {
+	in := new(HeaderSerializable)
+	if err := in.UnmarshalJSON(input); err != nil {
+		return err
+	}
+
+	updateToEth(in, eth)
+	updateToExtras(in, h)
+
+	return nil
+}
+
+//go:generate go run github.com/fjl/gencodec -type HeaderSerializable -field-override headerMarshaling -out gen_header_json.go
+//go:generate go run github.com/ava-labs/libevm/rlp/rlpgen -type HeaderSerializable -out gen_header_rlp.go
+
+// HeaderSerializable defines the header of a block in the Ethereum blockchain,
+// as it is to be serialized into RLP and JSON.
+type HeaderSerializable struct {
+	ParentHash  common.Hash    `json:"parentHash"       gencodec:"required"`
+	UncleHash   common.Hash    `json:"sha3Uncles"       gencodec:"required"`
+	Coinbase    common.Address `json:"miner"            gencodec:"required"`
+	Root        common.Hash    `json:"stateRoot"        gencodec:"required"`
+	TxHash      common.Hash    `json:"transactionsRoot" gencodec:"required"`
+	ReceiptHash common.Hash    `json:"receiptsRoot"     gencodec:"required"`
+	Bloom       Bloom          `json:"logsBloom"        gencodec:"required"`
+	Difficulty  *big.Int       `json:"difficulty"       gencodec:"required"`
+	Number      *big.Int       `json:"number"           gencodec:"required"`
+	GasLimit    uint64         `json:"gasLimit"         gencodec:"required"`
+	GasUsed     uint64         `json:"gasUsed"          gencodec:"required"`
+	Time        uint64         `json:"timestamp"        gencodec:"required"`
+	Extra       []byte         `json:"extraData"        gencodec:"required"`
+	MixDigest   common.Hash    `json:"mixHash"`
+	Nonce       BlockNonce     `json:"nonce"`
+	ExtDataHash common.Hash    `json:"extDataHash"      gencodec:"required"`
+
+	// BaseFee was added by EIP-1559 and is ignored in legacy headers.
+	BaseFee *big.Int `json:"baseFeePerGas" rlp:"optional"`
+
+	// ExtDataGasUsed was added by Apricot Phase 4 and is ignored in legacy
+	// headers.
+	//
+	// It is not a uint64 like GasLimit or GasUsed because it is not possible to
+	// correctly encode this field optionally with uint64.
+	ExtDataGasUsed *big.Int `json:"extDataGasUsed" rlp:"optional"`
+
+	// BlockGasCost was added by Apricot Phase 4 and is ignored in legacy
+	// headers.
+	BlockGasCost *big.Int `json:"blockGasCost" rlp:"optional"`
+
+	// BlobGasUsed was added by EIP-4844 and is ignored in legacy headers.
+	BlobGasUsed *uint64 `json:"blobGasUsed" rlp:"optional"`
+
+	// ExcessBlobGas was added by EIP-4844 and is ignored in legacy headers.
+	ExcessBlobGas *uint64 `json:"excessBlobGas" rlp:"optional"`
+
+	// ParentBeaconRoot was added by EIP-4788 and is ignored in legacy headers.
+	ParentBeaconRoot *common.Hash `json:"parentBeaconBlockRoot" rlp:"optional"`
+}
+
+// field type overrides for gencodec
```

Comment:
nit it would be great to expand on the comment since this has a sort of special role and may be thought of as unneeded by a developer

Reply:
this is a geth comment, leave it as is.